orttraining/orttraining/training_api/onnxruntime_training_c_api.cc

```diff
@@ -0,0 +1,226 @@
+                  _Outptr_ OrtTrainingSession** out) {
+  API_IMPL_BEGIN
+  std::unique_ptr<onnxruntime::training::api::TrainingSession> train_sess;
+  OrtStatus* status = nullptr;
+  *out = nullptr;
+
+  ORT_TRY {
+      train_sess = std::make_unique<onnxruntime::training::api::TrainingSession>(
+        options == nullptr ? onnxruntime::SessionOptions() : options->value,
+        env->GetEnvironment());
+
+      *out = reinterpret_cast<OrtTrainingSession*>(train_sess.release());
+  }
+  ORT_CATCH(const std::exception& e) {
+    ORT_HANDLE_EXCEPTION([&]() {
+      status = OrtApis::CreateStatus(ORT_FAIL, e.what());
+    });
+  }
+
+  return status;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtApis::InitializeTrainingSession, _Inout_ OrtTrainingSession* session,
+                    _Inout_ OrtCheckpointState* checkpoint_state,
+                    _In_ const ORTCHAR_T* train_model_path, _In_ const ORTCHAR_T* eval_model_path,
+                    _In_ const ORTCHAR_T* optimizer_model_path) {
+  API_IMPL_BEGIN
+
+  auto train_sess = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(session);
+  auto chkpt_state = reinterpret_cast<onnxruntime::training::api::CheckpointState*>(checkpoint_state);
+  ORT_API_RETURN_IF_STATUS_NOT_OK(train_sess->Initialize(chkpt_state->module_checkpoint_state.named_parameters,
+                                                         train_model_path,
+                                                         eval_model_path ? std::optional<std::string>{eval_model_path}
+                                                                         : std::nullopt,
+                                                         optimizer_model_path ? std::optional<std::string>{optimizer_model_path}
+                                                                              : std::nullopt));
+
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtApis::ResetGrad, _Inout_ OrtTrainingSession* session) {
+  API_IMPL_BEGIN
+  auto train_session = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(session);
+  ORT_API_RETURN_IF_STATUS_NOT_OK(train_session->ResetGrad());
+
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtApis::TrainStep, _Inout_ OrtTrainingSession* sess, _In_opt_ const OrtRunOptions* run_options,
+                  _In_reads_(inputs_len) const OrtValue* const* inputs, size_t inputs_len,
+                  size_t outputs_len, _Inout_updates_all_(outputs_len) OrtValue** outputs) {
+
+  API_IMPL_BEGIN
+  auto session = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(sess);
+  constexpr int queue_id = 0;
+
+  std::vector<OrtValue> feeds(inputs_len);
+
+  for (size_t i = 0; i != inputs_len; ++i) {
+    auto& ort_value = feeds[i] = *reinterpret_cast<const ::OrtValue*>(inputs[i]);
+    if (ort_value.Fence()) {
+      ort_value.Fence()->BeforeUsingAsInput(onnxruntime::kCpuExecutionProvider, queue_id);
```

Comment:
Should this be `kCpuExecutionProvider` or should the provider be extracted from the session options/ session before using fence?